### PR TITLE
[jaeger-v2] Refactor Configuration For Query Service

### LINKED ISSUE #5996: [jaeger-v2] Align Query Service Config With OTEL
### Requirement

In this issue, we want to align the configuration for the query service with OTEL (part of https://github.com/jaegertracing/jaeger/issues/5229)

### Tasks / outcomes

- [x] All config fields are tagged with mapstructure such that they can be addressable from YAML via "good" names
- [x] Configs implement Validate()
- [x] Configs via YAML support the same defaults as in jaeger-v1
- [x] Configs reuse standard elements of OTEL configs whenever possible, e.g. TLS, clientcfg, etc.
- [x] Configuration looks "logical" and properly grouped, not flattened with long weird names
- [x] We create a [migration guide document](https://docs.google.com/document/d/1ydTrEtIqJZxGb0RBkUAW-ZIaHNijo1pZSrn1TDSK-LU/edit) with a table that lists v1 CLI flags and the corresponding v2 config fields

### PR DESCRIPTION
**Pull request by mahadzaryab1**
## Which problem is this PR solving?
- Part of #5996

## Description of the changes
- Reworked the configuration for `jaeger-query` to have more logical groupings.
- [Migration document](https://docs.google.com/document/d/1-01QRgMFWbHGXXJqs5cL4kA-P10BfU0TD9Kvp3rEw3M/edit?usp=sharing)

## How was this change tested?
- CI

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: cmd/jaeger/internal/extension/jaegerquery/config.go [hunk 1]
```diff
@@ -18,12 +18,20 @@ var _ component.ConfigValidator = (*Config)(nil)
 // Config represents the configuration for jaeger-query,
 type Config struct {
 	queryApp.QueryOptionsBase `mapstructure:",squash"`
```
Comment (mahadzaryab1) [inline cmd/jaeger/internal/extension/jaegerquery/config.go:20]: @yurishkuro Do we want/need to rework [this configuration](https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/flags.go#L60-L78)? A few things I was hoping to get some guidance on: 
- Not all fields have mapstructure tags - are there any here that we want to add tags to? 
- For the fields that don't have mapstructure tags, should we move them to `QueryOptions`? 
- We have `Tenancy` twice

File: cmd/jaeger/internal/extension/jaegerquery/config.go [hunk 2]
```diff
@@ -18,12 +18,20 @@ var _ component.ConfigValidator = (*Config)(nil)
...
-	TraceStoragePrimary       string                  `valid:"required" mapstructure:"trace_storage"`
-	TraceStorageArchive       string                  `valid:"optional" mapstructure:"trace_storage_archive"`
-	MetricStorage             string                  `valid:"optional" mapstructure:"metric_storage"`
-	HTTP                      confighttp.ServerConfig `mapstructure:",squash"`
-	GRPC                      configgrpc.ServerConfig `mapstructure:",squash"`
-	Tenancy                   tenancy.Options         `mapstructure:"multi_tenancy"`
+	Connection                Connection `mapstructure:"connection"`
+	Storage                   Storage    `mapstructure:"storage"`
```
Comment (mahadzaryab1) [inline cmd/jaeger/internal/extension/jaegerquery/config.go:22]: @yurishkuro let me know if you have any feedback on the way i've grouped things here

File: cmd/jaeger/internal/extension/jaegerquery/factory.go [hunk 3]
```diff
@@ -27,13 +28,18 @@ func NewFactory() extension.Factory {
 
 func createDefaultConfig() component.Config {
 	return &Config{
-		HTTP: confighttp.ServerConfig{
-			Endpoint: ports.PortToHostPort(ports.QueryHTTP),
+		OptionsBase: app.OptionsBase{
+			BasePath: "/",
```
Comment (mahadzaryab1) [inline cmd/jaeger/internal/extension/jaegerquery/factory.go:32]: this was added to match the default of v1

File: cmd/query/app/flags.go [hunk 4]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
-	Path string `valid:"optional" mapstructure:"path"`
-	// LogAccess tells static handler to log access to static assets, useful in debugging
-	LogAccess bool `valid:"optional" mapstructure:"log_access"`
+type UI struct {
+	// ConfigFile is the path to a configuration file for the UI.
+	ConfigFile string `mapstructure:"config_file" valid:"optional"`
+	// AssetsPath is the path for the static assets for the UI (https://github.com/uber/jaeger-ui).
+	AssetsPath string `mapstructure:"assets_path" valid:"optional" `
+	// LogAccess tells static handler to log access to static assets, useful in debugging.
+	LogAccess bool `mapstructure:"log_access" valid:"optional"`
```
Comment (mahadzaryab1) [inline cmd/query/app/flags.go:57]: i made some changes here to group all ui configs - let me know if you have any thoughts

File: cmd/query/app/flags.go [hunk 5]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
-	// UIConfig is the path to a configuration file for the UI
-	UIConfig string `valid:"optional" mapstructure:"ui_config"`
-	// BearerTokenPropagation activate/deactivate bearer token propagation to storage
-	BearerTokenPropagation bool
-	// AdditionalHeaders
-	AdditionalHeaders http.Header
-	// MaxClockSkewAdjust is the maximum duration by which jaeger-query will adjust a span
-	MaxClockSkewAdjust time.Duration
-	// Tenancy configures tenancy for query
-	Tenancy tenancy.Options
```
Comment (mahadzaryab1) [inline cmd/query/app/flags.go:75]: moved to `Options` since the v2 config already has this

File: cmd/query/app/flags.go [hunk 6]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
-	BasePath string
-
-	StaticAssets QueryOptionsStaticAssets `valid:"optional" mapstructure:"static_assets"`
-
-	// UIConfig is the path to a configuration file for the UI
-	UIConfig string `valid:"optional" mapstructure:"ui_config"`
-	// BearerTokenPropagation activate/deactivate bearer token propagation to storage
-	BearerTokenPropagation bool
-	// AdditionalHeaders
-	AdditionalHeaders http.Header
```
Comment (mahadzaryab1) [inline cmd/query/app/flags.go:71]: moved to Options since the v2 config already has this in the HTTP config

File: cmd/query/app/flags.go [hunk 7]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
+	// BearerTokenPropagation activate/deactivate bearer token propagation to storage.
+	BearerTokenPropagation bool `mapstructure:"bearer_token_propagation"`
+	// MaxClockSkewAdjust is the maximum duration by which jaeger-query will adjust a span.
+	MaxClockSkewAdjust time.Duration `mapstructure:"max_clock_skew_adjust"`
 	// EnableTracing determines whether traces will be emitted by jaeger-query.
-	EnableTracing bool
+	EnableTracing bool `mapstructure:"enable_tracing"`
 }
 
-// QueryOptions holds configuration for query service
```
Comment (yurishkuro) [inline cmd/query/app/flags.go:80]: why change the name of the struct? It introduces unnecessary changes elsewhere.

File: cmd/query/app/flags.go [hunk 8]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
-type QueryOptions struct {
-	QueryOptionsBase
+// Options holds configuration for query service
+type Options struct {
+	OptionsBase
 
+	// Tenancy configures tenancy for query
+	Tenancy tenancy.Options
+	// AdditionalHeaders
+	AdditionalHeaders http.Header
```
Comment (yurishkuro) [inline cmd/query/app/flags.go:81]: I don't like this. The objective should be to align v1 and v2 configs more, but you're going in the opposite direction - repeating more things in v2 which are done differently from v1. So far the only difference in v2 config were the storage names (which are legitimately different from v1 as a mechanism), and the HTTP/GRPC settings because we wanted to reuse OTEL-standard structs for those. I don't recall why tenancy was separate in v2 and not part of the base. But the ideal situation would be that v2 config is just storage names + Base struct, nothing else.

File: cmd/query/app/flags.go (hunk 7 above)
Comment (mahadzaryab1) [reply cmd/query/app/flags.go:80]: In most of the other configs we just had `Options` and I thought prefixing with `Query` was redundant since the file is in the `query` package. But I can revert this change to make the PR smaller.

File: cmd/query/app/flags.go (hunk 8 above)
Comment (mahadzaryab1) [reply cmd/query/app/flags.go:81]: Ah okay got it - sorry about that! Let me move these back to the BaseConfig.

File: cmd/query/app/flags.go [hunk 9]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
-	EnableTracing bool
+	EnableTracing bool `mapstructure:"enable_tracing"`
 }
 
 // QueryOptions holds configuration for query service
 type QueryOptions struct {
 	QueryOptionsBase
 
+	// AdditionalHeaders
+	AdditionalHeaders http.Header
```
Comment (mahadzaryab1) [inline cmd/query/app/flags.go:81]: @yurishkuro I moved tenancy to the base config. What're your thoughts on `AdditionalHeaders`. The following s field exists in OTEL's `confighttp.ServerConfig`.
```go
// Additional headers attached to each HTTP response sent to the client.
// Header values are opaque since they may be sensitive.
ResponseHeaders map[string]configopaque.String `mapstructure:"response_headers"`
```
As a result of this, should we leave `AdditionalHeaders` in the base config or should we have it be specific to v1?

File: cmd/query/app/static_handler.go [hunk 10]
```diff
@@ -36,12 +36,12 @@ var (
...
 func RegisterStaticHandler(r *mux.Router, logger *zap.Logger, qOpts *QueryOptions, qCapabilities querysvc.StorageCapabilities) io.Closer {
-	staticHandler, err := NewStaticAssetsHandler(qOpts.StaticAssets.Path, StaticAssetsHandlerOptions{
+	staticHandler, err := NewStaticAssetsHandler(qOpts.UI.AssetsPath, StaticAssetsHandlerOptions{
 		BasePath:            qOpts.BasePath,
-		UIConfigPath:        qOpts.UIConfig,
+		UIConfigPath:        qOpts.UI.ConfigFile,
 		StorageCapabilities: qCapabilities,
 		Logger:              logger,
-		LogAccess:           qOpts.StaticAssets.LogAccess,
+		LogAccess:           qOpts.UI.LogAccess,
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/query/app/static_handler.go:44]: also, is the first argument `qOpts.UI.AssetsPath` actually required or can also be empty? If so I would move everything into `StaticAssetsHandlerOptions`

File: cmd/query/app/flags.go [hunk 11]
```diff
@@ -48,39 +48,37 @@ var tlsHTTPFlagsConfig = tlscfg.ServerFlagsConfig{
...
-	// AdditionalHeaders
-	AdditionalHeaders http.Header
-	// MaxClockSkewAdjust is the maximum duration by which jaeger-query will adjust a span
-	MaxClockSkewAdjust time.Duration
-	// Tenancy configures tenancy for query
-	Tenancy tenancy.Options
+	// BasePath is the base path for all HTTP routes.
+	BasePath string `mapstructure:"base_path"`
+	// UI contains configuration related to the Jaeger UI.
+	UI UI `mapstructure:"ui" valid:"optional"`
```
[the remark this answers is not in the document]
Comment (mahadzaryab1) [reply cmd/query/app/flags.go:65]: each of the fields in the struct are marked as optional so I don't think its needed there - removed it

File: cmd/query/app/static_handler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mahadzaryab1) [reply cmd/query/app/static_handler.go:44]: @yurishkuro `AssetsPath` is also optional. Do we want to replace the `UIConfig` struct with `StaticAssetsHandlerOptions`. Or do we want make `UIConfig` part of the `StaticAssetsHandlerOptions`?

File: cmd/query/app/flags.go (hunk 9 above)
Comment (mahadzaryab1) [reply cmd/query/app/flags.go:81]: that's a good point! let me try seeing if i can do this

File: cmd/query/app/static_handler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/query/app/static_handler.go]: I meant:
```
type StaticAssetsHandlerOptions struct {
	// BasePath            string
	// UIConfigPath        string
	// LogAccess           bool
	UICongig
}
```
Why do we need to move packages?

File: cmd/query/app/flags.go (hunk 9 above)
Comment (mahadzaryab1) [reply cmd/query/app/flags.go:81]: @yurishkuro I looked into this and it should be possible but the number of changes will be large. Should we make the changes in this PR or should we do it in a following one?

Comment (yurishkuro) [reply cmd/query/app/flags.go:81]: yeah, we can do in another PR, this one is already heavy

File: cmd/query/app/static_handler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mahadzaryab1) [reply cmd/query/app/static_handler.go]: @yurishkuro do we want to move `BasePath` to `UIConfig` as well? It's used in the query server to set up the routes for the API handler so I wasn't sure if that config is UI specific.

[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/query/app/static_handler.go]: no, base path is not a UI config, it affects everything including API endpoints

File: cmd/jaeger/internal/extension/jaegerquery/factory.go [hunk 12]
```diff
@@ -27,13 +28,18 @@ func NewFactory() extension.Factory {
 
 func createDefaultConfig() component.Config {
 	return &Config{
-		HTTP: confighttp.ServerConfig{
-			Endpoint: ports.PortToHostPort(ports.QueryHTTP),
+		QueryOptionsBase: app.QueryOptionsBase{
+			BasePath: "/",
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/factory.go:32]: this looks new

File: cmd/jaeger/internal/extension/jaegerquery/server.go [hunk 13]
```diff
@@ -83,7 +83,7 @@ func (s *server) Start(_ context.Context, host component.Host) error {
 		return err
 	}
 
-	tm := tenancy.NewManager(&s.config.Tenancy)
+	tm := tenancy.NewManager(&s.config.QueryOptionsBase.Tenancy)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/server.go:86]: since the struct is embedded, I don't think you need `QueryOptionsBase.` qualifier

File: cmd/jaeger/internal/extension/jaegerquery/config.go [hunk 14]
```diff
@@ -10,20 +10,31 @@ import (
...
+}
+
+type Connection struct {
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerquery/config.go:24]: I am not completely sold on this - it's a different pattern from what most OTEL receivers are using, so why do we want this extra level of indirection?

File: cmd/query/app/static_handler.go [hunk 15]
```diff
@@ -62,9 +64,8 @@ type StaticAssetsHandler struct {
 
 // StaticAssetsHandlerOptions defines options for NewStaticAssetsHandler
 type StaticAssetsHandlerOptions struct {
+	UIConfig            UIConfig
```
Comment (yurishkuro) [inline cmd/query/app/static_handler.go:67]: I would embed it
```suggestion
	UIConfig
```

File: cmd/jaeger/internal/extension/jaegerquery/factory.go (hunk 12 above)
Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/factory.go:32]: @yurishkuro I added this in to match the default of v1 from https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/flags.go#L99. But it actually looks like that's being done in https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/static_handler.go#L126-L128 anyway. I'll remove it.

File: cmd/jaeger/internal/extension/jaegerquery/config.go (hunk 14 above)
Comment (mahadzaryab1) [reply cmd/jaeger/internal/extension/jaegerquery/config.go]: @yurishkuro When the `tenancy` field was still in this config, I thought grouping by connection made sense but it's not really needed anymore. I moved `HTTP` and `GRPC` to `Config`.

File: cmd/query/app/flags_test.go [hunk 16]
```diff
@@ -35,9 +35,9 @@ func TestQueryBuilderFlags(t *testing.T) {
 	})
 	qOpts, err := new(QueryOptions).InitFromViper(v, zap.NewNop())
 	require.NoError(t, err)
-	assert.Equal(t, "/dev/null", qOpts.StaticAssets.Path)
-	assert.True(t, qOpts.StaticAssets.LogAccess)
-	assert.Equal(t, "some.json", qOpts.UIConfig)
+	assert.Equal(t, "/dev/null", qOpts.UIConfig.AssetsPath)
+	assert.True(t, qOpts.UIConfig.LogAccess)
+	assert.Equal(t, "some.json", qOpts.UIConfig.ConfigFile)
```
Comment (yurishkuro) [inline cmd/query/app/flags_test.go:40]: UIConfig prefix not needed

File: cmd/query/app/static_handler.go [hunk 17]
```diff
@@ -36,12 +36,14 @@ var (
 
 // RegisterStaticHandler adds handler for static assets to the router.
 func RegisterStaticHandler(r *mux.Router, logger *zap.Logger, qOpts *QueryOptions, qCapabilities querysvc.StorageCapabilities) io.Closer {
-	staticHandler, err := NewStaticAssetsHandler(qOpts.StaticAssets.Path, StaticAssetsHandlerOptions{
+	staticHandler, err := NewStaticAssetsHandler(qOpts.UIConfig.AssetsPath, StaticAssetsHandlerOptions{
+		UIConfig: UIConfig{
+			ConfigFile: qOpts.UIConfig.ConfigFile,
+			LogAccess:  qOpts.UIConfig.LogAccess,
+		},
```
Comment (yurishkuro) [inline cmd/query/app/static_handler.go:43]:
```suggestion
		UIConfig: qOpts.UIConfig,
```

**Review by yurishkuro — COMMENTED** [review]
Couple cleanups, otherwise lgtm

File: cmd/query/app/flags_test.go (hunk 16 above)
Comment (mahadzaryab1) [reply cmd/query/app/flags_test.go:40]: @yurishkuro its needed here because `UIConfig` is not embedded in `QueryOptions` - its only embedded in `StaticAssetsHandlerOptions`

**Review by yurishkuro — APPROVED** [review]
Thanks! 🎉
